src/PhpDoc/PhpDocPrivateClassConstFetchFinder.php

```diff
@@ -0,0 +1,41 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\PhpDoc;
+
+use Nette\Utils\Strings;
+use PHPStan\Analyser\Scope;
+use PHPStan\Node\Constant\ClassConstantFetch;
+use PHPStan\PhpDocParser\Ast\NodeTraverser;
+use PHPStan\ShouldNotHappenException;
+
+class PhpDocPrivateClassConstFetchFinder
+{
+
+	public function __construct(
+		private PhpDocStringResolver $phpDocStringResolver,
+	)
+	{
+	}
+
+	/**
+	 * @return list<ClassConstantFetch>
+	 */
+	public function findClassConstFetches(string $phpDoc, Scope $scope): array
+	{
+		if (!$scope->isInClass()) {
+			throw new ShouldNotHappenException();
+		}
+
+		if (Strings::match($phpDoc, '(\bself\s*::)i') === null) {
+			return [];
+		}
+
+		$visitor = new PhpDocPrivateClassConstFetchFinderVisitor($scope->getClassReflection(), $scope);
+
+		$phpDocNode = $this->phpDocStringResolver->resolve($phpDoc);
```

Comment:
I really like this, it's almost perfect. What gives me a little pause is that all PHPDocs are now parsed multiple times which would lead to performance problems. The correct way to access them is through FileTypeMapper. Also you shouldn't have to do any regexes, but correctly resolve the class name with NameScope.

Of course the issue is that the thing returned by FileTypeMapper is ResolvedPhpDocBlock. Which is about accessing the useful PHPDoc things like ParamTag and many others.

One way to solve this would be to add `getClassConstFetches` on ResolvedPhpDocBlock. But the class is a bit tricky to work with - because it can be created multiple ways (create, createEmpty, merge), we need to be careful every time a new property is added.